File: src/main/java/com/telegram/codex/integration/telegram/application/webhook/TelegramCommandHandler.java
package com.telegram.codex.integration.telegram.application.webhook;

import com.telegram.codex.conversation.application.session.SessionService;
import com.telegram.codex.conversation.domain.MessageConstants;
import com.telegram.codex.integration.telegram.domain.InboundMessage;
import org.springframework.stereotype.Component;

import java.util.Optional;

@Component
public class TelegramCommandHandler {

    private final TelegramCommandRegistry commandRegistry;
    private final SessionService sessionService;
    private final TelegramStatusMessageBuilder messageBuilder;
    private final TelegramCommandResponder responder;
    private final CompactCommandExecutor compactCommandExecutor;

    public TelegramCommandHandler(
        TelegramCommandRegistry commandRegistry,
        SessionService sessionService,
        TelegramStatusMessageBuilder messageBuilder,
        TelegramCommandResponder responder,
        CompactCommandExecutor compactCommandExecutor
    ) {
        this.commandRegistry = commandRegistry;
        this.sessionService = sessionService;
        this.messageBuilder = messageBuilder;
        this.responder = responder;
        this.compactCommandExecutor = compactCommandExecutor;
    }

    public boolean handle(InboundMessage message) {
        Optional<TelegramCommandRegistry.TelegramCommand> command = commandRegistry.resolve(message);
        if (command.isEmpty()) {
            return false;
        }
        execute(command.get(), message);
        return true;
    }

    private void execute(TelegramCommandRegistry.TelegramCommand command, InboundMessage message) {
        switch (command) {
            case START -> {
                sessionService.reset(message.chatId());
                responder.reply(message, MessageConstants.START_MESSAGE);
            }
            case NEW_SESSION -> {
                sessionService.reset(message.chatId());
                responder.reply(message, MessageConstants.NEW_SESSION_MESSAGE);
            }
            case HELP -> responder.reply(message, MessageConstants.HELP_MESSAGE);
            case STATUS -> responder.reply(message, messageBuilder.buildStatusMessage(message.chatId()));
            case SESSION -> responder.reply(message, messageBuilder.buildSessionMessage(message.chatId()));
            case MEMORY -> responder.reply(message, messageBuilder.buildMemoryMessage(message.chatId()));
            case FORGET -> {
                sessionService.resetMemory(message.chatId());
                responder.reply(message, MessageConstants.RESET_MEMORY_MESSAGE);
            }
            case COMPACT -> compactCommandExecutor.execute(message);
        }
    }
}

File: src/telegram/commands/telegram_command_handler.rs
use std::sync::Arc;

use anyhow::Result;

use crate::conversation::session::session_service::SessionService;
use crate::telegram::commands::compact_command_executor::CompactCommandExecutor;
use crate::telegram::commands::telegram_command_registry::{TelegramCommand, TelegramCommandRegistry};
use crate::telegram::commands::telegram_command_responder::TelegramCommandResponder;
use crate::telegram::commands::telegram_status_message_builder::TelegramStatusMessageBuilder;
use crate::telegram::shared::inbound_message::InboundMessage;
use crate::telegram::shared::message_constants;

pub struct TelegramCommandHandler {
    command_registry: Arc<TelegramCommandRegistry>,
    session_service: Arc<SessionService>,
    message_builder: Arc<TelegramStatusMessageBuilder>,
    responder: Arc<TelegramCommandResponder>,
    compact_command_executor: Arc<CompactCommandExecutor>,
}

impl TelegramCommandHandler {
    pub fn new(
        command_registry: Arc<TelegramCommandRegistry>,
        session_service: Arc<SessionService>,
        message_builder: Arc<TelegramStatusMessageBuilder>,
        responder: Arc<TelegramCommandResponder>,
        compact_command_executor: Arc<CompactCommandExecutor>,
    ) -> Self {
        Self {
            command_registry,
            session_service,
            message_builder,
            responder,
            compact_command_executor,
        }
    }

    /// Returns `true` when the message was a command and has been answered.
    pub async fn handle(&self, message: &InboundMessage) -> Result<bool> {
        let Some(command) = self.command_registry.resolve(message) else {
            return Ok(false);
        };
        match command {
            TelegramCommand::Start => {
                self.session_service.reset(&message.chat_id).await?;
                self.responder.reply(message, &message_constants::START_MESSAGE).await?;
            }
            TelegramCommand::NewSession => {
                self.session_service.reset(&message.chat_id).await?;
                self.responder.reply(message, message_constants::NEW_SESSION_MESSAGE).await?;
            }
            TelegramCommand::Help => {
                self.responder.reply(message, &message_constants::HELP_MESSAGE).await?;
            }
            TelegramCommand::Status => {
                let text = self.message_builder.build_status_message(&message.chat_id).await?;
                self.responder.reply(message, &text).await?;
            }
            TelegramCommand::Session => {
                let text = self.message_builder.build_session_message(&message.chat_id).await?;
                self.responder.reply(message, &text).await?;
            }
            TelegramCommand::Memory => {
                let text = self.message_builder.build_memory_message(&message.chat_id).await?;
                self.responder.reply(message, &text).await?;
            }
            TelegramCommand::Forget => {
                self.session_service.reset_memory(&message.chat_id).await?;
                self.responder.reply(message, message_constants::RESET_MEMORY_MESSAGE).await?;
            }
            TelegramCommand::Compact => {
                self.compact_command_executor.execute(message).await?;
            }
        }
        Ok(true)
    }
}
